### DISCUSSION TITLE
Usability of the vsomeip library

### DISCUSSION THREAD

**Discussion by nachtfuchs** [COVESA/vsomeip#557] [Q&A]
Hi,

tl;dr: **What do you guys do to get a better understanding of the vsomeip interface and its capabilities**?

I am noticing that I am having troubles to use the library in my context. I run the library on a Linux machine and use it to communicate with an embedded device. The embedded device uses BroadR-Reach as a cable and some unknown SOME/IP stack. I know that the latter works, because I have a home grown reference SOME/IP implementation on my Linux machine. I want to exchange the current " home grown reference SOME/IP implementation" with vsomeip. I have been working on it for a couple of weeks already, and I notice that it is almost impossible for me to gain development momentum. 
I feel that on of the issues is that I am lacking an understanding of the "correct" usage of the json-file. I simply have trouble to understand some of the parameters beyond the examples. Also, some timing values have no units, i.e. "service-discovery" -> "initial_delay_min" (is it seconds, milliseconds or whatever?).
Another issue is that some of the service parameters are very well hidden. Examples are the major and minor versions of a service. This comes hand in hand with the order of methods that need to be called for a vsomeip::application to make it run correctly. Whenever I encounter problems with my setup I go back to the examples and transfer their methods to my particular challenge. Problem is that it just doesn't work, although it should work.
All things combined, I am circling around the same problem for a long time. 

What I have done so far is, I read the:
- wiki vsomeip in 10 Minutes (to understand the "interface")
- vsomeip/build/documentation/vsomeipUserGuide.html (to understand the json-file)

I have also built the examples and played around with them successfully in a single, and two node setups. I changed the parameters in the json-files, and also played with the source files. I thought I understood the library. When it comes to my challenge, I see that I do not understand it.
